app/GioForegroundService.java

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Unlicense OR MIT
+
+package org.gioui;
+import android.app.Notification;
+import android.app.Service;
+import android.app.Notification;
+import android.app.Notification.Builder;
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.ComponentName;
+import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.os.IBinder;
+import android.os.Bundle;
+
+// GioForegroundService implements a Service required to use the FOREGROUND_SERVICE
+// permission on Android, in order to run an application in the background.
+// See https://developer.android.com/guide/components/foreground-services for
+// more details. To add this permission to your application, import
+// gioui.org/app/permission/foreground and use the Start method from that
+// package to control this service.
+public class GioForegroundService extends Service {
+	private String channelName;
+
+	@Override public int onStartCommand(Intent intent, int flags, int startId) {
+		// Get the channel parameters from intent extras and package metadata.
+		Bundle extras = intent.getExtras();
+		String title = extras.getString("title");
+		String text = extras.getString("text");
+		Context ctx = getApplicationContext();
+		try {
+			ComponentName svc = new ComponentName(this, this.getClass());
+			Bundle metadata = getPackageManager().getServiceInfo(svc, PackageManager.GET_META_DATA).metaData;
+			if (metadata != null) {
+				channelName = metadata.getString("org.gioui.ForegroundChannelName");
+				String channelDesc = metadata.getString("org.gioui.ForegroundChannelDesc");
+				String channelID = metadata.getString("org.gioui.ForegroundChannelID");
+				int notificationID = metadata.getInt("org.gioui.ForegroundNotificationID", 0x42424242);
+				this.createNotificationChannel(channelDesc, channelID, channelName);
+				Intent launchIntent = getPackageManager().getLaunchIntentForPackage(ctx.getPackageName());
+				PendingIntent pending = PendingIntent.getActivity(ctx, notificationID, launchIntent, Intent.FLAG_ACTIVITY_CLEAR_TASK);
+				Notification.Builder builder = new Notification.Builder(ctx, channelID)
```

Comment:
The channel constructor is only available in API 26 and later.

Reply:
https://support.google.com/googleplay/android-developer/answer/9859152?visit_id=637402646062706504-1797308153&rd=1#targetsdk&zippy=%2Ctarget-api-level-requirements-for-play-console

wontfix.

Reply:
Note the difference between `target` SDK and `minimum` SDK. It's true that our target must be at least 30, and that `gogio` now requires SDK 31 or later to build.

However, our `minimum` SDK is [16](https://git.sr.ht/~eliasnaur/gio/tree/main/item/cmd/gogio/androidbuild.go#L362) which means that Android code should be prepared to deal with missing API.